File: src/main/java/jp/kentan/minecraft/neko_event/gacha/GachaManager.java
package jp.kentan.minecraft.neko_event.gacha;

import jp.kentan.minecraft.neko_event.NekoEvent;
import jp.kentan.minecraft.neko_event.config.listener.ConfigListener;
import jp.kentan.minecraft.neko_event.config.provider.GachaConfigProvider;
import jp.kentan.minecraft.neko_event.config.provider.SignConfigProvider;
import jp.kentan.minecraft.neko_event.gacha.model.Gacha;
import jp.kentan.minecraft.neko_event.listener.SignEventListener;
import jp.kentan.minecraft.neko_event.listener.SignListener;
import jp.kentan.minecraft.neko_event.module.key.KeyManager;
import jp.kentan.minecraft.neko_event.ticket.TicketProvider;
import jp.kentan.minecraft.neko_event.util.Log;
import jp.kentan.minecraft.neko_event.util.NekoUtil;
import org.apache.commons.lang.StringUtils;
import org.bukkit.*;
import org.bukkit.block.Sign;
import org.bukkit.command.CommandSender;
import org.bukkit.command.ConsoleCommandSender;
import org.bukkit.entity.Player;
import org.bukkit.event.block.Action;
import org.bukkit.event.block.SignChangeEvent;
import org.bukkit.event.player.PlayerInteractEvent;
import org.bukkit.plugin.java.JavaPlugin;

import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.Map;

public class GachaManager implements SignListener, ConfigListener<Gacha> {

    private final static GachaManager sInstance = new GachaManager();

    private final static Server SERVER = Bukkit.getServer();
    private final static ConsoleCommandSender COMMAND_SENDER = SERVER.getConsoleSender();

    private static Map<String, Gacha> sGachaMap = new HashMap<>();

    public static void setup(JavaPlugin plugin){
        GachaConfigProvider.bindListener(sInstance);
        GachaConfigProvider.load();

        SignEventListener.bindGachaSignListener(sInstance);

        plugin.getCommand("gacha").setExecutor(new GachaCommandExecutor());
    }

    private static void play(Player player, Gacha gacha){
        Gacha.Component component = gacha.getByRandom();

        final Location location = player.getLocation();
        final Sound soundEffect;

        if(component.hasCommands()){ //あたり
            final String playerName = player.getName();

            component.getCommands().forEach(cmd -> SERVER.dispatchCommand(COMMAND_SENDER, cmd.replace("{player}", playerName)));

            if(gacha.hasGetMessage()) {
                player.sendMessage(NekoEvent.PREFIX + gacha.getMessage().replace("{name}", component.getName()));
            }

            if(gacha.hasBroadcastMessage()){
                final String broadcastMsg = gacha.getBroadcastMessage().replace("{player}", player.getName()).replace("{name}", component.getName());
                NekoUtil.broadcastMessage(NekoEvent.PREFIX + broadcastMsg, player);
            }

            Log.info(player.getName() + "にｶﾞﾁｬ(" + gacha.getId() + ")で" + component.getName() + ChatColor.RESET + "を与えました.");

            soundEffect = Sound.ENTITY_PLAYER_LEVELUP;
        }else{ //はずれ
            if(gacha.hasMissMessage()) {
                player.sendMessage(NekoEvent.PREFIX + gacha.getMissMessage());
            }

            soundEffect = Sound.ENTITY_PIG_DEATH;
        }

        if(gacha.enableEffect()){
            player.playSound(location, soundEffect, 1.0f, 0.0f);
        }
    }

    public static void play(Player player, String gachaId){
        Gacha gacha = sGachaMap.get(gachaId);
        if(gacha == null){
            Log.error(GACHA_ID_NOT_FOUND.replace("{id}", gachaId));
            return;
        }

        play(player, gacha);
    }

    static void play(String playerName, String gachaId, String ticketCost){
        Player player = NekoUtil.toPlayer(playerName);

        if(player == null){
            return;
        }

        Gacha gacha = sGachaMap.get(gachaId);
        if(gacha == null){
            Log.error(GACHA_ID_NOT_FOUND.replace("{id}", gachaId));
            return;
        }

        if(ticketCost == null || TicketProvider.remove(player, NekoUtil.toInteger(ticketCost), gacha.isRequireVoteTicket())) {
            play(player, gacha);
        }
    }

    private static void play(Player player, String gachaId, int ticketCost){
        Gacha gacha = sGachaMap.get(gachaId);
        if(gacha == null){
            Log.error(GACHA_ID_NOT_FOUND.replace("{id}", gachaId));
            return;
        }

        if(ticketCost > 0 && !TicketProvider.remove(player, ticketCost, gacha.isRequireVoteTicket())) {
            return;
        }

        play(player, gacha);
    }

    private static void play(Player player, String gachaId, String keyId){
        if(!KeyManager.removeFormInventory(player, keyId)) {
            return;
        }

        Gacha gacha = sGachaMap.get(gachaId);
        if(gacha == null){
            Log.error(GACHA_ID_NOT_FOUND.replace("{id}", gachaId));
            return;
        }

        play(player, gacha);
    }

    static void sendList(CommandSender sender){
        StringBuilder builder = new StringBuilder(GACHA_ID_LIST_TEXT + "\n");

        sGachaMap.entrySet().stream()
                .sorted(Map.Entry.comparingByKey())
                .forEach(e -> {
                            builder.append(" ");
                            builder.append(e.getKey());
                            builder.append(": ");
                            builder.append(e.getValue().getName());
                            builder.append('\n');
                            builder.append(ChatColor.RESET);
                        }
                );

        sender.sendMessage(builder.toString());
    }

    static void sendInfo(CommandSender sender, String gachaId){
        final Gacha gacha = sGachaMap.get(gachaId);

        if(gacha == null){
            sender.sendMessage(NekoEvent.PREFIX + ChatColor.YELLOW + GACHA_ID_NOT_FOUND.replace("{id}",gachaId));
            return;
        }

        sender.sendMessage(GACHA_INFO_TEXT);
        sender.sendMessage(" ID: " + gachaId);
        sender.sendMessage(" 名前: " + gacha.getName());
        sender.sendMessage(" 要素数: " + gacha.getSize());
        sender.sendMessage(" ｹﾞｯﾄﾒｯｾｰｼﾞ: " + gacha.hasGetMessage());
        sender.sendMessage(" ﾐｽﾒｯｾｰｼﾞ: " + gacha.hasMissMessage());
        sender.sendMessage(" ﾌﾞﾛｰﾄﾞｷｬｽﾄ: " + gacha.hasBroadcastMessage());
        sender.sendMessage(" ｻｳﾝﾄﾞｴﾌｪｸﾄ: " + gacha.enableEffect());
    }

    private static boolean hasGachaMetadata(Sign sign){
        Location location = sign.getLocation();

        if(!SignConfigProvider.hasMetadata(location, GACHA_ID_METADATA_KEY)){
            Log.error("看板" + NekoUtil.toString(location) + "にﾒﾀﾃﾞｰﾀ(gachaId)が不足しています.");
            return false;
        }

        boolean hasTicketCost = SignConfigProvider.hasMetadata(location, GACHA_TICKET_COST_METADATA_KEY);
        boolean hasKeyId = SignConfigProvider.hasMetadata(location, GACHA_KEY_ID_METADATA_KEY);

        if(!hasTicketCost && !hasKeyId){
            Log.error("看板" + NekoUtil.toString(location) + "にﾒﾀﾃﾞｰﾀ(gachaTicketCost)または(gachaKeyId)が不足しています.");
            return false;
        }

        return true;
    }

    public static boolean hasGacha(String id){
        return sGachaMap.containsKey(id);
    }

    @Override
    public void onSignChanged(SignChangeEvent event) {
        final Player player = event.getPlayer();
        final String gachaId = event.getLine(2);
        final String strCostData = event.getLine(3);

        int ticketCost = 0;
        String keyId = null;
        if(strCostData != null && strCostData.length() > 0){
            if(StringUtils.isNumeric(strCostData)){
                ticketCost = NekoUtil.toInteger(strCostData);
            }else{
                keyId = strCostData;
            }
        }

        final Gacha gacha = sGachaMap.get(gachaId);

        if(gacha == null){
            player.sendMessage(NekoEvent.PREFIX + ChatColor.YELLOW + GACHA_ID_NOT_FOUND.replace("{id}", gachaId));
            return;
        }

        final Sign sign = (Sign)event.getBlock().getState();

        //看板のメタデータにガチャ情報を保存
        Map<String, Object> metadataMap = new LinkedHashMap<>();
        metadataMap.put(GACHA_ID_METADATA_KEY, gachaId);
        metadataMap.put(GACHA_TICKET_COST_METADATA_KEY, ticketCost);

        if(keyId != null) {
            metadataMap.put(GACHA_KEY_ID_METADATA_KEY, keyId);
        }

        if(!SignConfigProvider.saveMetadata(sign.getLocation(), metadataMap)){
            Log.error("看板ﾒﾀﾃﾞｰﾀの保存に失敗しました.");
            return;
        }

        event.setLine(0, GACHA_SIGN_INDEX);
        event.setLine(1, gacha.getName());

        if(keyId == null) { //Event ticket
            event.setLine(2, "");

            if(ticketCost > 0) {
                event.setLine(3, GACHA_SIGN_TICKET_COST.replace("{amount}", strCostData));
            }else{
                event.setLine(3, GACHA_SIGN_TICKET_COST_FREE);
            }
        }else{ //key item
            event.setLine(2, GACHA_SIGN_KEY_COST.replace("{amount}", Integer.toString(KeyManager.getKeyAmount(keyId))));

            final String keyName = KeyManager.getKeyName(keyId);
            event.setLine(3, (keyName != null) ? keyName : keyId);
        }
    }

    @Override
    public void onPlayerInteract(PlayerInteractEvent event, Sign sign) {
        if(event.getAction() != Action.RIGHT_CLICK_BLOCK || !hasGachaMetadata(sign)){
            return;
        }

        final String gachaId = (String) SignConfigProvider.getMetadata(sign.getLocation(), GACHA_ID_METADATA_KEY);

        final int ticketCost = (int) SignConfigProvider.getMetadata(sign.getLocation(), GACHA_TICKET_COST_METADATA_KEY);
        final String keyId = (String) SignConfigProvider.getMetadata(sign.getLocation(), GACHA_KEY_ID_METADATA_KEY);

        if(!sGachaMap.containsKey(gachaId)){
            Log.error(GACHA_ID_NOT_FOUND.replace("{id}",gachaId));
            return;
        }

        if(keyId == null) { //Event ticket
            play(event.getPlayer(), gachaId, ticketCost);
        }else{ //Key item
            play(event.getPlayer(), gachaId, keyId);
        }
    }

    @Override
    public void onUpdate(Map<String, Gacha> gachaMap) {
        sGachaMap.clear();
        sGachaMap.putAll(gachaMap);
    }

    private final static String GACHA_ID_METADATA_KEY = "gachaId";
    private final static String GACHA_TICKET_COST_METADATA_KEY = "gachaTicketCost";
    private final static String GACHA_KEY_ID_METADATA_KEY = "gachaKeyId";

    public final static String GACHA_SIGN_INDEX = ChatColor.translateAlternateColorCodes('&', "&8&l[&d&lガチャ&8&l]");
    private final static String GACHA_SIGN_PLAY = ChatColor.translateAlternateColorCodes('&', "&9&n1プレイ");
    private final static String GACHA_SIGN_TICKET_COST = GACHA_SIGN_PLAY + ChatColor.translateAlternateColorCodes('&', "&r &a&n{amount}枚");
    private final static String GACHA_SIGN_KEY_COST = GACHA_SIGN_PLAY + ChatColor.translateAlternateColorCodes('&', "&r &a&n{amount}個");
    private final static String GACHA_SIGN_TICKET_COST_FREE = ChatColor.translateAlternateColorCodes('&', "&c無料");

    private final static String GACHA_ID_LIST_TEXT = ChatColor.translateAlternateColorCodes('&', "&7--------- &dガチャ一覧 &7---------&r");
    private final static String GACHA_INFO_TEXT = ChatColor.translateAlternateColorCodes('&', "&7--------- &dガチャ情報 &7---------");

    public final static String GACHA_ID_NOT_FOUND = "ガチャID({id})は存在しません.";
}

File: src/main/java/jp/kentan/minecraft/nekoevent/manager/GachaManager.kt
package jp.kentan.minecraft.nekoevent.manager

import jp.kentan.minecraft.nekoevent.NekoEvent
import jp.kentan.minecraft.nekoevent.component.GachaCost
import jp.kentan.minecraft.nekoevent.component.GachaCost.*
import jp.kentan.minecraft.nekoevent.component.TicketType
import jp.kentan.minecraft.nekoevent.component.model.Gacha
import jp.kentan.minecraft.nekoevent.config.ConfigUpdateListener
import jp.kentan.minecraft.nekoevent.config.provider.GachaConfigProvider
import jp.kentan.minecraft.nekoevent.config.provider.SignConfigProvider
import jp.kentan.minecraft.nekoevent.listener.SignListener
import jp.kentan.minecraft.nekoevent.util.*
import org.bukkit.Bukkit
import org.bukkit.ChatColor
import org.bukkit.Sound
import org.bukkit.block.Sign
import org.bukkit.command.CommandSender
import org.bukkit.entity.Player
import org.bukkit.event.block.Action
import org.bukkit.event.block.SignChangeEvent
import org.bukkit.event.player.PlayerInteractEvent

class GachaManager(
        private val ticketManager: TicketManager,
        private val keyManager: KeyManager,
        private val gachaConfig: GachaConfigProvider,
        private val signConfig: SignConfigProvider
) : ConfigUpdateListener<Gacha>, SignListener {

    companion object {
        private val SIGN_INDEX = "&8&l[&d&lガチャ&8&l]".formatColorCode()
        val SIGN_KEY = Pair("[gacha]", SIGN_INDEX)

        private const val ID_METADATA_KEY = "gachaId"
        private const val COST_METADATA_KEY = "gachaCost"
        private const val COST_DETAIL_METADATA_KEY = "gachaCostDetail"
    }

    private val gachaMap = mutableMapOf<String, Gacha>()

    init {
        gachaConfig.listener = this
    }

    fun getGachaIdList() = gachaMap.values.map { it.id }.sorted()

    fun play(strPlayer: String, gachaId: String) {
        val player = strPlayer.toPlayerOrError() ?: return

        play(player, gachaId)
    }

    fun play(player: Player, gachaId: String) {
        val gacha = gachaMap.getOrError(gachaId) ?: return

        play(player, gacha)
    }

    private fun playWithTicket(player: Player, gachaId: String, ticket: TicketType, amount: Int) {
        if (ticketManager.take(player, ticket, amount)) {
            play(player, gachaId)
        }
    }

    private fun playWithKey(player: Player, gachaId: String, keyId: String) {
        if (keyManager.use(player, keyId)) {
            play(player, gachaId)
        }
    }

    private fun play(player: Player, gacha: Gacha) {
        val component = gacha.getByRandom()

        val soundEffect: Sound

        if (component.isWin()) {
            component.commandList.forEach { Bukkit.getServer().dispatchCommand(Bukkit.getConsoleSender(), it.replace("{player}", player.name)) }

            gacha.winMessage?.let {
                player.sendMessage(NekoEvent.PREFIX + it.replace("{name}", component.name))
            }

            gacha.broadcastMessage?.let {
                if (component.isDisabledBroadcast) {
                    return@let
                }

                val message = it
                        .replace("{player}", player.name)
                        .replace("{username}", player.displayName)
                        .replace("{name}", component.name)

                player.broadcastMessageWithoutMe(NekoEvent.PREFIX + message)
            }

            soundEffect = Sound.ENTITY_PLAYER_LEVELUP

            Log.info(player.name + "にｶﾞﾁｬ(" + gacha.id + ")で" + component.name + ChatColor.RESET + "を与えました.")
        } else {
            gacha.loseMessage?.let {
                player.sendMessage(NekoEvent.PREFIX + it)
            }

            soundEffect = Sound.ENTITY_PIG_DEATH
        }

        if (gacha.enabledEffect) {
            player.playSound(player.location, soundEffect, 1.0f, 0.0f)
        }
    }

    fun demo(sender: CommandSender, gachaId: String, strTimes: String) {
        val gacha = gachaMap.getOrError(gachaId) ?: return

        val time = strTimes.toIntOrNull() ?: let {
            Log.error("$strTimes は0以上の整数にしてください.")
            return
        }

        if (time > 1000000) {
            sender.sendMessage(NekoEvent.PREFIX + ChatColor.YELLOW + "100万回以上は指定出来ません.")
            return
        }

        val resultMap = mutableMapOf<String, Int>()

        for (i in 0 until time) {
            val key = gacha.getByRandom().name
            val count = resultMap[key]

            if (count != null) {
                resultMap.replace(key, count + 1)
            } else {
                resultMap[key] = 1
            }
        }

        sender.sendMessage(NekoEvent.PREFIX + "ｶﾞﾁｬ(" + gachaId + ")を" + time + "回引いた結果")

        resultMap.forEach { name, count -> sender.sendMessage(name + ChatColor.RESET + ": " + count + "回") }
    }

    fun sendList(sender: CommandSender) {
        val sb = StringBuilder("&7--------- &dガチャ一覧 &7---------&r".formatColorCode())
        sb.append('\n')

        gachaMap.toSortedMap().forEach { id, gacha ->
            sb.append(' ')
            sb.append(id)
            sb.append(": ")
            sb.append(gacha.formatName)
            sb.append('\n')
            sb.append(ChatColor.RESET)
        }

        sender.sendMessage(sb.toString())
    }

    fun sendInfo(sender: CommandSender, gachaId: String) {
        val gacha = gachaMap.getOrError(gachaId) ?: return

        val messages = arrayOf(
                "&7--------- &dガチャ情報 &7---------".formatColorCode(),
                " ID: $gachaId",
                " 名前: ${gacha.formatName}",
                " コンポ数: ${gacha.componentList.size}",
                " あたり: ${gacha.winMessage}",
                " はずれ: ${gacha.loseMessage}",
                " ﾌﾞﾛｰﾄﾞｷｬｽﾄ: ${gacha.broadcastMessage}",
                " エフェクト: ${gacha.enabledEffect}")

        sender.sendMessage(messages)
    }

    fun reload() {
        gachaConfig.load()
    }

    override fun onConfigUpdate(dataMap: Map<String, Gacha>) {
        gachaMap.clear()
        gachaMap.putAll(dataMap)

        Log.info("${dataMap.size}個のガチャを読み込みました.")
    }

    /**
     * 看板フォーマット
     * 0: [gacha]
     * 1: gachaId
     * 2: [jp.kentan.minecraft.nekoevent.model.GachaCost].id
     * 3: cost detail (ex. ticket num)
     */
    override fun onSignChanged(event: SignChangeEvent) {
        val gachaId = event.getLine(1)
        val strCost = event.getLine(2)
        val strCostDetail = event.getLine(3)

        // コスト判定
        val cost = GachaCost.find(strCost) ?: let {
            Log.error("ガチャコスト($strCost)は不正な値です.")
            return
        }

        val signMetadataMap = LinkedHashMap<String, Any>().apply {
            put(ID_METADATA_KEY, gachaId)
            put(COST_METADATA_KEY, cost.name)
        }

        val costDetailText: String

        when (cost) {
            EVENT_TICKET, VOTE_TICKET -> {
                val amount = strCostDetail.toIntOrError() ?: return

                costDetailText = "&9&n1プレイ&r &a&n${amount}枚".formatColorCode()

                signMetadataMap[COST_DETAIL_METADATA_KEY] = amount
            }
            KEY -> {
                val key = keyManager.getKey(strCostDetail) ?: let {
                    Log.error("鍵($strCost)は存在しません.")
                    return
                }

                costDetailText = key.name

                signMetadataMap[COST_DETAIL_METADATA_KEY] = key.id
            }
            FREE -> {
                costDetailText = "&c無料".formatColorCode()
            }
        }


        val gacha = gachaMap.getOrError(gachaId) ?: return

        val sign = event.block.state as Sign
        if (!signConfig.save(sign.location, signMetadataMap)) {
            Log.error("看板ﾒﾀﾃﾞｰﾀの保存に失敗しました.")
            return
        }


        event.setLine(0, SIGN_INDEX)
        event.setLine(1, gacha.formatName)
        event.setLine(2, cost.signText)
        event.setLine(3, costDetailText)
    }

    override fun onPlayerInteract(event: PlayerInteractEvent, sign: Sign) {
        if (event.action != Action.RIGHT_CLICK_BLOCK) {
            return
        }

        val gachaId = signConfig.getMetadata(sign.location, ID_METADATA_KEY) as String? ?: return

        val strCost = signConfig.getMetadata(sign.location, COST_METADATA_KEY) as String? ?: return
        val cost = GachaCost.valueOf(strCost)
        val costDetail = signConfig.getMetadata(sign.location, COST_DETAIL_METADATA_KEY)?.toString() ?: return

        when (cost) {
            FREE -> play(event.player, gachaId)
            EVENT_TICKET -> playWithTicket(event.player, gachaId, TicketType.EVENT, costDetail.toIntOrNull() ?: 0)
            VOTE_TICKET  -> playWithTicket(event.player, gachaId, TicketType.VOTE, costDetail.toIntOrNull() ?: 0)
            KEY -> playWithKey(event.player, gachaId, costDetail)
        }
    }

    private fun MutableMap<String, Gacha>.getOrError(id: String): Gacha? {
        return gachaMap[id] ?: let {
            Log.error("ガチャ($id)は存在しません.")
            return@let null
        }
    }
}